core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala

```diff
@@ -399,7 +426,14 @@ private[storage] class BlockInfoManager extends Logging {
     try {
       val wrapper = new BlockInfoWrapper(newBlockInfo, lock)
       while (true) {
-        val previous = blockInfoWrappers.putIfAbsent(blockId, wrapper)
+        val previous = invisibleRDDBlocks.synchronized {
+          val res = blockInfoWrappers.putIfAbsent(blockId, wrapper)
+          if (res == null && trackingCacheVisibility) {
+            // Added to invisible blocks if it doesn't exist before.
+            blockId.asRDDId.foreach(invisibleRDDBlocks.add)
```

Comment:
I'd suggest updating `invisibleRDDBlocks` within `lockForWriting()` when the write locking is acquired.

Reply:
Currently the way we check whether a block is visible is to check the information from both `blockInfoWrappers` and `invisibleRDDBlocks`, so we will need the synchronized block to modify the state to avoid concurrent issues (`invisibleRDDBlocks` should be updated once new item put into `blockInfoWrappers` within the synchronized block). It will be easier to have the syncrhonized block here.

```
  private[spark] def isRDDBlockVisible(blockId: RDDBlockId): Boolean = {
    if (trackingCacheVisibility) {
      invisibleRDDBlocks.synchronized {
        blockInfoWrappers.containsKey(blockId) && !invisibleRDDBlocks.contains(blockId)
      }
    } else {
      // Always be visible if the feature flag is disabled.
      true
    }
  }
```

Reply:
Currently we syncrhonized all the write operations, and also the read operations which need to check both the variables. For other read operations, since the variabled won't change, I think synchronized block for such read operations may be not necessary.